src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Windows.cs

```diff
@@ -288,6 +289,30 @@ unsafe long GetFileLengthCore()
 
                 if (!Interop.Kernel32.GetFileInformationByHandleEx(this, Interop.Kernel32.FileStandardInfo, &info, (uint)sizeof(Interop.Kernel32.FILE_STANDARD_INFO)))
                 {
+                    if (this._path?.StartsWith(@"\\?\") ?? false)
+                    {
+                        byte[] buffer = ArrayPool<byte>.Shared.Rent(4096);
+                        bool success = Interop.Kernel32.DeviceIoControl(
+                            this,
+                            dwIoControlCode: Interop.Kernel32.IOCTL_STORAGE_READ_CAPACITY,
+                            lpInBuffer: IntPtr.Zero,
+                            nInBufferSize: 0,
+                            lpOutBuffer: buffer,
+                            nOutBufferSize: 4096,
+                            out _,
+                            IntPtr.Zero);
+
+                        if (!success)
+                        {
+                            throw Win32Marshal.GetExceptionForLastWin32Error(Path);
+                        }
+
+                        Span<byte> bufferSpan = new(buffer);
+                        success = MemoryMarshal.TryRead(bufferSpan, out Interop.Kernel32.StorageReadCapacity storageReadCapacity);
+                        Debug.Assert(success);
+                        return storageReadCapacity.DiskLength;
```

Comment:
You could use a smaller array length here; the `StorageReadCapacity` struct is only 28 bytes long.

Even better the P/Invoke could accept a `Span<byte>` (I think it can done with the `LibraryImport` marshalling and we could entirely avoid the array allocation, but that's not necessarily something that needs to be done now.

You can also use `MemoryMarshal.Read` to get the struct from the bytes; the buffer will always be big enough.

Reply:
Updated.

For some reason setting `nOutBufferSize` in [DeviceIoControl](https://docs.microsoft.com/en-us/windows/win32/api/ioapiset/nf-ioapiset-deviceiocontrol) to 28 results in ERROR_INSUFFICIENT_BUFFER. It has to be 32.

`TryRead` as well as much of this code came from here to have consistency.

https://github.com/dotnet/runtime/blob/a5f3676cc71e176084f0f7f1f6beeecd86fbeafc/src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs#L605

I don't mind either way. 

My demo project for this issue uses CsWin32 code generator, which makes this call extremely clean

https://github.com/virzak/retrieve-disks/blob/44d822f14cf260af518b1b7a39c5c4992456c9da/src/RetrieveDisks/DiskIO.cs#L155-L169